VMware: Improve vmware_vmkernel

ckotte opened the pull request:
##### SUMMARY
* Update description and examples
  * It's not necessary to describe that the VMkernel devices are called 'vnic' internally
  * Especially because a VM nic is called 'vnic' as well
* Add check mode support
* Show settings that were updated / will be updated in the verbose message
* Add default gateway configuration (Check box: 'Override default gateway for this adapter')
* Remove unused option `vlan_id`
* Add vDS support
  * Create port on vDS (Replaces PR https://github.com/ansible/ansible/pull/45636; Fixes https://github.com/ansible/ansible/issues/43351 & #50989)
  * Migrate port from vSS to vDS, or vice versa (Replaces PR https://github.com/ansible/ansible/pull/45636)
  * I didn't get a response from the (other) PR owner and I needed to implement vmkernel ports on vDS in my environment. I also think that we should use `vswitch_name` and `dvswitch_name` and `device_name` should be optional instead of using `is_dvs_switch` and make `device_name` a required option.
* Add TCP/IP stack configuration (fixes #48706)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
vmware_vmkernel

##### ANSIBLE VERSION
```
ansible 2.6.3
  config file = /root/ansible-vmware/ansible.cfg
  configured module search path = [u'/root/ansible-vmware/library']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Jul 13 2018, 13:06:57) [GCC 4.8.5 20150623 (Red Hat 4.8.5-28)]
```

##### ADDITIONAL INFORMATION
Before:
```
TASK [esxi : test] ***************************************************************************************************************************************************************************************************************************
changed: [esxi_host -> localhost] => {"changed": true, "result": "vmk0"}
```
After:
```
TASK [esxi : test] ***************************************************************************************************************************************************************************************************************************
changed: [esxi_host -> localhost] => {"changed": true, "changed_msg": "VMkernel Adapter 'vmk0' updated: MTU (9000) and vMotion (False)", "device": "vmk0"}
```

ckotte committed:
Improve vmware_vmkernel module

* update description and examples
* show updated settings
* add check mode support
* Remove unused option `vlan_id`
* add vDS support
* add TCP/IP stack support

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vmkernel.py:
```diff
@@ -374,90 +333,188 @@ def host_vmk_update(self):
-        ip_spec = vim.host.IpConfig()
-        if self.network_type == 'dhcp':
-            ip_spec.dhcp = True
+        changed = False
```
```suggestion
        changed = changed_settings = changed_services = changed_service_vmotion = changed_service_mgmt = changed_service_ft = changed_service_vsan = changed_service_prov = changed_service_rep = changed_service_rep_nfc = False
```

ckotte replied:
done

Akasurde commented on lib/ansible/modules/cloud/vmware/vmware_vmkernel.py:
```diff
@@ -374,90 +333,188 @@ def host_vmk_update(self):
+            if changed_services:
+                vnic_manager = self.esxi_host_obj.configManager.virtualNicManager
+
+                if changed_service_vmotion and self.enable_vmotion:
```
```suggestion
                if changed_service_vmotion:
                   operation = 'select' if self.enable_vmotion else 'deselect'
                   self.set_service_type(vnic_manager=vnic_manager, vmk=self.vnic, service_type='vmotion',
                                          operation=operation)
                        
```

Akasurde replied:
same for below.

ckotte replied:
Looks much cleaner. Thx.

ckotte commented:
@Akasurde @dagwieers 
> ```
> lib/ansible/modules/cloud/vmware/vmware_vmkernel.py:0:0: E324 Value for "default" from the argument_spec ({'type': 'static', 'device': None, 'ip_address': None, 'subnet_mask': None, 'default_gateway': None}) for "network" does not match the documentation ({'type': 'static', 'device': 'None', 'ip_address': 'None', 'subnet_mask': 'None', 'default_gateway': 'None'})
> ```

Any idea how to fix this documentation error?

Akasurde commented:
@mattclay I added `vmware_vmkernel` is ignore list with 324 as suboption validation is not supported right now. Let me know if it is ok, so that I can merge this PR. Thanks.

ckotte commented:
@Akasurde 
> I added `vmware_vmkernel` in ignore list with 324 as suboption validation is not supported right now. Let me know if it is ok, so that I can merge this PR. Thanks.

I fixed the documentation error. Looks like you cannot use default=None since the documentation doesn't accept it.

Akasurde commented:
@pdellaert @dericcrago @Im0 @tchernomax @jeking3 @Tomrrow9 Could you please review this ?

Akasurde reviewed (approved):
LGTM

dandunckelman commented:
@ckotte I just tried using this, but I'm getting this error:

    Portgroup 'pg1' not found on vDS 'dvs1'

If I add that VMK thru the vCenter UI, then it reports what I would expect:

    VMkernel Adapter already configured properly

Your tests don't cover using `dvswitch_name`, which I assume is a limitation of the vcsim.

I believe this issue is already described in #50989.

Please let me know if there's more info I can supply to help.

ckotte commented:
> Please let me know if there's more info I can supply to help.

@dandunckelman I need more information to troubleshoot this.

I just execute a function with the switch object and the portgroup name. The switch was obviously found, but the function cannot find a valid portgroup.

```
self.dv_switch_obj = find_dvs_by_name(self.content, self.vds_name)
if not self.dv_switch_obj:
    module.fail_json(msg="vDS '%s' not found" % self.vds_name)
self.port_group_obj = find_dvspg_by_name(self.dv_switch_obj, self.port_group_name)
if not self.port_group_obj:
    module.fail_json(msg="Portgroup '%s' not found on vDS '%s'" % (self.port_group_name, self.vds_name))
```

```
def find_dvspg_by_name(dv_switch, portgroup_name):

    portgroups = dv_switch.portgroup

    for pg in portgroups:
        if pg.name == portgroup_name:
            return pg

    return None
```

gundalow commented:
Marking as WIP till the above is worked out (which will remove `label/shipit`)

dandunckelman commented:
@ckotte correct. As described in #50989, this appears to be the way ESX operates when it comes to VDS.

I ran the same code you use against my vcenter, and the portgroup that I wanted to add was in the list. But when run directly against the esx host, it only showed portgroups that were already connected. As a result, the `find_dvspg_by_name` method will always `return None` when trying to create a VMK with a portgroup that isn't already on the host.

Even running `esxcli network vswitch dvs vmware list` only lists the connected VDS and the connected portgroups.

I wish I could find documentation to confirm my findings, but I haven't had any luck.

ckotte commented:
> But when run directly against the esx host, it only showed portgroups that were already connected.

What do you mean with "already connected distributed portgroup"? You can only see distributed portgroups with connected VMs or VMkernel adapters?

I think you have to authenticate to a vCenter and not to an ESXi host if you want to create a new VMkernel on a distributed portgroup. Otherwise, we would need to add a new parameter vcenter that's only required if the connection is local and if a DVS is used. However, I would rather add a comment in the module notes that a vCenter connection is necessary if a DVS is involved or add code that aborts the execution if it's a local ESXi connection.

dandunckelman commented:
"already connected" meaning a VMK already added to the host.

I think you have to add a param similar to what's done in https://docs.ansible.com/ansible/latest/modules/vmware_dvs_host_module.html#vmware-dvs-host-module

Because I need to be able to state which host receives the VMK while using my vcenter to find the necessary portgroup. So, AFAIK, you'd need both:
- `esxi_hostname` for the host receiving the VMK
- `hostname` for the hostname or IP address of the vcenter

ckotte commented:
Why not just connecting to the vCenter? The DVS must be configured via vCenter anyway.BTW. I would also need to pass the vCenter credentials since I need to authenticate to query the portgroups.Instead of doing this in the module, this should be done in the playbook.However, I need to change the examples because they use esxi_hostname when using a DVS... I will also add a note to the description...
-------- Original message --------From: Dan Dunckel <notifications@github.com> Date: 2/13/19  15:23  (GMT+01:00) To: ansible/ansible <ansible@noreply.github.com> Cc: Christian Kotte <christian.kotte@gmx.de>, Mention <mention@noreply.github.com> Subject: Re: [ansible/ansible] [WIP] VMware: Improve vmware_vmkernel (#47270) "already connected" meaning a VMK already added to the host.
I think you have to add a param similar to what's done in https://docs.ansible.com/ansible/latest/modules/vmware_dvs_host_module.html#vmware-dvs-host-module
Because I need to be able to state which host receives the VMK while using my vcenter to find the necessary portgroup. So, AFAIK, you'd need both:

esxi_hostname for the host receiving the VMK
hostname for the hostname or IP address of the vcenter


—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/ansible/ansible","title":"ansible/ansible","subtitle":"GitHub repository","main_image_url":"https://github.githubassets.com/images/email/message_cards/header.png","avatar_image_url":"https://github.githubassets.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/ansible/ansible"}},"updates":{"snippets":[{"icon":"PERSON","message":"@dandunckelman in #47270: \"already connected\" meaning a VMK already added to the host.\r\n\r\nI think you have to add a param similar to what's done in https://docs.ansible.com/ansible/latest/modules/vmware_dvs_host_module.html#vmware-dvs-host-module\r\n\r\nBecause I need to be able to state which host receives the VMK while using my vcenter to find the necessary portgroup. So, AFAIK, you'd need both:\r\n- `esxi_hostname` for the host receiving the VMK\r\n- `hostname` for the hostname or IP address of the vcenter"}],"action":{"name":"View Pull Request","url":"https://github.com/ansible/ansible/pull/47270#issuecomment-463216787"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/ansible/ansible/pull/47270#issuecomment-463216787",
"url": "https://github.com/ansible/ansible/pull/47270#issuecomment-463216787",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]

dandunckelman commented:
@ckotte If you have a working playbook that does just that, then please share.

I'm struggling to see how that could work.

ckotte commented:
Just like this:

```
- name: Configure vMotion VMkernel adapter
  vmware_vmkernel:
    hostname: "{{ vcenter_server if esxi_connect_via_vcenter else inventory_hostname }}"
    username: "{{ vcsa_username if esxi_connect_via_vcenter else esxi_username }}"
    password: "{{ vcsa_password if esxi_connect_via_vcenter else esxi_password }}"
    esxi_hostname: "{{ inventory_hostname }}"
    vswitch: "{{ dvswitch_vmotion }}"
    portgroup: "{{ dvportgroup_vmotion }}"
    mtu: "{{ esxi_mtu }}"
    enable_vmotion: True
    network:
      type: static
      ip_address: "{{ vmotion_ipv4 }}"
      subnet_mask: "{{ esxi_vmotion_network.netmask }}"
    state: present
    validate_certs: "{{ validate_vcenter_certs }}"
  ignore_errors: "{{ esxi_ignore_errors }}"
  delegate_to: localhost
```

You have to create proper group vars structure depending on your environment. For example, I grouped the ESXi hosts into clusters and I specified the vCenter server FQDN in this file. The credentials can be defined in the same file or on another level.

I usually authenticate via vCenter, but I can authenticate locally as well via an extra variable that I can override. However, this wouldn't work with new VMkernel adapters that need to connect to dvPortgroups.

dandunckelman commented:
@ckotte how would that work when [this PR's vmware_vmkernel module](https://github.com/ansible/ansible/blob/b2c144ee234a3b38a6319ff00e0355a1188bc8bc/lib/ansible/modules/cloud/vmware/vmware_vmkernel.py) never uses the `hostname` param?

> However, this wouldn't work with new VMkernel adapters that need to connect to dvPortgroups.

Is that the problem described in https://github.com/ansible/ansible/issues/43351?
- That issue best describes my problem
- The workaround listed in the issue description is *exactly* what I have to do ATM
- The [last comment](https://github.com/ansible/ansible/issues/43351#issuecomment-430625572) was you saying that this PR fixes it, but [as I described above](https://github.com/ansible/ansible/pull/47270#issuecomment-463197165):
  - When trying to add a VMK using a dvPortgroup to a host, the call to `find_dvspg_by_name` ONLY returns the list of dvPortgroups already added to the host via other VMKs.

I feel like we're talking on different wavelengths. Please let me know how I can clarify anything. I want this to work for us, but ATM, the bottom line is this: **this PR does not solve the problem described in https://github.com/ansible/ansible/issues/43351**.

ckotte commented:
> @ckotte how would that work when [this PR's vmware_vmkernel module](https://github.com/ansible/ansible/blob/b2c144ee234a3b38a6319ff00e0355a1188bc8bc/lib/ansible/modules/cloud/vmware/vmware_vmkernel.py) never uses the `hostname` param?

`hostname`, `username`, and `password` are used by every module to authenticate. They are inherited..

> * When trying to add a VMK using a dvPortgroup to a host, the call to `find_dvspg_by_name` ONLY returns the list of dvPortgroups already added to the host via other VMKs.

Yes. Already discussed with you many times. It cannot find all dvPortgroups if not connected to vCenter... You already said it works if connected to vCenter.

> **this PR does not solve the problem described in #43351**.

Of course it does. You just don't get the point. #43351 uses vCenter to authenticate as well. I can create VMkernel adapters if I authenticate to vCenter and NOT ESXi... I did that many times already...

> * The workaround listed in the issue description is _exactly_ what I have to do ATM

Then please use this workaround if it helps you and if don't want to authenticate to vCenter. Why do you need to pollute my PR with this??? Open a new issue if you still don't understand

dandunckelman commented:
@ckotte 

> `hostname`, `username`, and `password` are used by every module to authenticate. They are inherited..

This! 1000x this. I had NO idea.

<img src="https://media.giphy.com/media/5aLrlDiJPMPFS/giphy.gif" width=100 />

I was getting hung up on the fact that I wasn't seeing `hostname` in the code for the `vmware_vmkernel` module. A quick scan of the code brought me to [lib/ansible/module_utils/vmware_rest_client.py](https://github.com/ansible/ansible/blob/64a6dcdd1df64624083880b209f09548f4ec97b0/lib/ansible/module_utils/vmware_rest_client.py), which I believe does exactly what you're saying.

Regardless, I reworked my playbook to use `vmware_vmkernel` as you described, and everything worked as expected!

I feel SO stupid. I'm terribly sorry for burning a bunch of time on this, when I had a simple misunderstanding. Thanks for the PR and your patience in helping me see my mistake.

ckotte commented:
> I feel SO stupid. I'm terribly sorry for burning a bunch of time on this, when I had a simple misunderstanding. Thanks for the PR and your patience in helping me see my mistake.

No problem. I'm happy to help. I just got a bit annoyed at the end ;)

Akasurde commented:
@ckotte Ping me when you are done with your changes and we can merge this. Thanks, for the help.

ckotte commented:
@Akasurde I'm done. I just added a note.

ilfox commented:
When running the vmware_vmkernel module I'm still getting dvswitch is not supported and vlan_id is still required.

dandunckelman commented:
@ilfox this hasn't gone out w/ a release yet. I believe it's slated for 2.8.x.

If you need it, then you need to install from the `devel` branch.

Currently, my `requirements.txt` has this:

    -e git://github.com/ckotte/ansible.git@vmware_vmkernel#egg=ansible

but you could use this:

    -e git://github.com/ansible/ansible.git@devel#egg=ansible

ilfox commented:
Tested and worked on 6.7U1. Thank you!
